src/app/mesh/qgsmaptooleditmeshframe.cpp

```diff
@@ -2661,35 +2730,70 @@ void QgsMapToolEditMeshFrame::addVertex(
   double zValue;
   QgsPointXY effectivePoint = mapPoint;
 
-  if ( mCadDockWidget->cadEnabled() && mCurrentFaceIndex == -1 )
-    zValue = currentZValue();
-  else if ( mapPointMatch.isValid() && mapPointMatch.layer() && QgsWkbTypes::hasZ( mapPointMatch.layer()->wkbType() ) )
+  // initial Z value that can be modified
+  zValue = currentZValue();
+
+  bool isOnMesh = mCurrentFaceIndex != -1 || ( mCurrentEdge.first != -1 && mCurrentEdge.second != -1 );
+
+  if ( mWidgetActionDigitizing->zValueSourceType() == QgsMeshEditDigitizingAction::Terrain || ( mWidgetActionDigitizing->zValueSourceType() == QgsMeshEditDigitizingAction::InMeshUseMeshTerrainOtherwise && !isOnMesh ) )
   {
-    const QgsPoint layerPoint = mapPointMatch.interpolatedPoint( mCanvas->mapSettings().destinationCrs() );
-    zValue = layerPoint.z();
+    const QgsAbstractTerrainProvider *terrainProvider = QgsProject::instance()->elevationProperties()->terrainProvider();
+    const QgsCoordinateTransform transformation = QgsCoordinateTransform( mCurrentLayer->crs(), terrainProvider->crs(), QgsProject::instance() );
+
+    try
+    {
+      const QgsPointXY point = transformation.transform( effectivePoint.x(), effectivePoint.y() );
+      zValue = terrainProvider->heightAt( point.x(), point.y() );
+    }
+    catch ( const QgsCsException & )
+    {
+      zValue = std::numeric_limits<double>::quiet_NaN();
+    }
+
+    // either outside of terrain or the point cannot be transformed to terrainProvider CRS, use currentZValue
+    if ( std::isnan( zValue ) )
+    {
+      QgisApp::instance()->messageBar()->pushMessage(
+        tr( "Terrain Z Value" ),
+        tr( "Z Value from project terrain could not be obtained, setting default value %1." ).arg( mZValueWidget->getDefaultValue() ),
+        Qgis::MessageLevel::Warning,
+        3
```

Comment:
I'd leave this to the default timeout value.
```suggestion
      QgisApp::instance()->messageBar()->pushMessage(
        tr( "Terrain Z Value" ),
        tr( "Z Value from project terrain could not be obtained, setting default value %1." ).arg( mZValueWidget->getDefaultValue() ),
        Qgis::MessageLevel::Warning
```